src/Libraries/Microsoft.Extensions.Telemetry.Abstractions/Logging/LoggerMessageState.cs

```diff
@@ -0,0 +1,127 @@
+    /// Adds a property to the state.
+    /// </summary>
+    /// <param name="propertyName">The name of the property to add.</param>
+    /// <param name="propertyValue">The value of the property to add.</param>
+    /// <param name="classification">The data classification of the property value.</param>
+    /// <exception cref="ArgumentNullException"><paramref name="propertyName"/> is <see langword="null"/>.</exception>
+    /// <exception cref="ArgumentException"><paramref name="propertyName" /> is empty or contains exclusively whitespace,
+    /// or when a property of the same name has already been added.
+    /// </exception>
+    public void AddProperty(string propertyName, object? propertyValue, DataClassification classification)
+    {
+        _ = Throw.IfNull(propertyName);
+        ClassifiedProperties.Add(new(propertyName, propertyValue, classification));
+    }
+
+    /// <summary>
+    /// Resets state of this container as described in <see cref="IResettable.TryReset"/>.
+    /// </summary>
+    /// <returns>
+    /// <see langword="true" /> if the object successfully reset and can be reused.
+    /// </returns>
+    public bool TryReset()
+    {
+        Properties.Clear();
+        ClassifiedProperties.Clear();
+        _propertyCollector.ParameterName = string.Empty;
+        return true;
+    }
+
+#if NET6_0_OR_GREATER
+    /// <summary>
+    /// Gets log define options configured to skip the log level enablement check.
+    /// </summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static LogDefineOptions SkipEnabledCheckOptions { get; } = new() { SkipEnabledCheck = true };
+#endif
+
+    /// <summary>
+    /// Gets the list of properties added to this instance.
+    /// </summary>
+    [SuppressMessage("Design", "CA1002:Do not expose generic lists", Justification = "Not intended for application use")]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public List<KeyValuePair<string, object?>> Properties { get; } = new();
+
+    /// <summary>
+    /// Gets a list of properties which must receive redaction before being used.
+    /// </summary>
+    [SuppressMessage("Design", "CA1002:Do not expose generic lists", Justification = "Not intended for application use")]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public List<ClassifiedProperty> ClassifiedProperties { get; } = new();
+
+    /// <summary>
+    /// Gets the property collector instance.
+    /// </summary>
+    /// <param name="parameterName">The name of the parameter to prefix in front of all property names inserted into the collector.</param>
+    /// <returns>The collector instance.</returns>
+    /// <remarks>
+    /// This method is used by the logger message code generator to get an instance of a collector to
+    /// use when invoking a custom property collector method.
+    /// </remarks>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public ILogPropertyCollector GetPropertyCollector(string parameterName)
```

Comment:
The way you suggest is how it was previously (see LogMethodHelper as a reference). While looking at it, I realized this pattern was wasting cycles in the 99% case.

ILogPropertyCollector is rarely used, say in < 1% of log statements. It's necessarily an interface since it's implemented by customers and making a base class instead would be constraining. In the case of ILogPropertyCollector, we need to inject the parameter name in front of all the properties. Rather than paying the cost of an if check any time a property is added to the property bag, this cost is now borne only by the ILogPropertyCollector case, instead of burdening the other 99% of cases.

Reply:
> It's necessarily an interface since it's implemented by customers

Is there an example of where a customer would implement ILogPropertyCollector? I see where customers would call it when using the LogProperties attribute, but I didn't see there was a scenario where customers would implement a different implementation. This is just for my understanding, I'm not worried about it.